piker/brokers/kucoin.py

```diff
@@ -0,0 +1,670 @@
+            ping_interval: int = data['instanceServers'][0]['pingInterval']
+            return data['token'], ping_interval
+        elif data:
+            log.error(
+                f'Error making request for Kucoin ws token -> {data.json()["msg"]}'
+            )
+
+    async def _get_pairs(
+        self,
+    ) -> dict[str, KucoinMktPair]:
+        entries = await self._request('GET', '/symbols')
+        syms = {kucoin_sym_to_fqsn(item['name']): KucoinMktPair(**item) for item in entries}
+
+        log.info('Kucoin market pairs fetched')
+        return syms
+
+    async def cache_pairs(
+        self,
+        # normalize: bool = True,
+    ) -> dict[str, KucoinMktPair]:
+        '''
+        Get cached pairs and convert keyed symbols into fqsns if ya want
+
+        '''
+        if not self._pairs:
+            self._pairs = await self._get_pairs()
+
+        return self._pairs
+
+    async def search_symbols(
+        self,
+        pattern: str,
+        limit: int = 30,
+    ) -> dict[str, KucoinMktPair]:
+        data = await self._get_pairs()
+
+        matches = fuzzy.extractBests(
+            pattern, data, score_cutoff=35, limit=limit)
+        # repack in dict form
+        return {item[0].name: item[0] for item in matches}
+
+    async def last_trades(self, sym: str) -> list[AccountTrade]:
+        trades = await self._request('GET', f'/accounts/ledgers?currency={sym}', 'v1')
+        trades = AccountResponse(**trades)
+        return trades.items
+
+    async def _get_bars(
+        self,
+        fqsn: str,
+        start_dt: Optional[datetime] = None,
+        end_dt: Optional[datetime] = None,
+        limit: int = 1000,
+        as_np: bool = True,
+        type: str = '1min',
+    ) -> np.ndarray:
+        '''
+        Get OHLC data and convert to numpy array for perffff:
+        https://docs.kucoin.com/#get-klines
+
+        '''
+        # Generate generic end and start time if values not passed
+        # Currently gives us 12hrs of data
+        if end_dt is None:
+            end_dt = pendulum.now('UTC').add(minutes=1)
+
+        if start_dt is None:
+            start_dt = end_dt.start_of('minute').subtract(minutes=limit)
+
+        start_dt = int(start_dt.timestamp())
+        end_dt = int(end_dt.timestamp())
+
+        kucoin_sym = fqsn_to_kucoin_sym(fqsn, self._pairs)
+
+        url = f'/market/candles?type={type}&symbol={kucoin_sym}&startAt={start_dt}&endAt={end_dt}'
+        bars = []
+
+        for i in range(10):
+            data = await self._request(
+                'GET',
+                url,
+                api_v='v1',
+            )
+
+            if not isinstance(data, list):
+                # Do a gradual backoff if Kucoin is rate limiting us
+                backoff_interval = i
+                log.warn(
+                    f'History call failed, backing off for {backoff_interval}s')
+                await trio.sleep(backoff_interval)
+            else:
+                bars = data
```

Comment:
considering you're overriding `bars: list[list]` each backoff algo case, seems weird you pre-define it above as an empty list.

maybe just make the pre-def a type annot only?